src/messaging/messaging.ts

```diff
@@ -245,16 +311,26 @@ export class Messaging implements MessagingInterface {
       throw new FirebaseMessagingError(
         MessagingClientErrorCode.INVALID_ARGUMENT, 'dryRun must be a boolean');
     }
-    return this.getUrlPath()
-      .then((urlPath) => {
-        const request: { message: Message; validate_only?: boolean } = { message: copy };
-        if (dryRun) {
-          request.validate_only = true;
-        }
-        return this.messagingRequestHandler.invokeRequestHandler(FCM_SEND_HOST, urlPath, request);
+
+    const clientMessage = this.convertToClientMessage(copy);
+
+    const client = this.getFcmServiceClient();
+
+    return client.getProjectId()
```

Comment:
Is this looking up the project ID from `FcmServiceClient`? Then we must pass `projectId` to the `FcmServiceClient` constructor to avoid having GAPIC do extra work.

Reply:
Made the adjustment, using `utils.findProjectId()`